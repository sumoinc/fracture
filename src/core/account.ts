import { Component } from "projen";
import { Organization } from "./organization";
import { OrganizationalUnit } from "./organizational-unit";
import { Region, RegionOptions } from "./region";

export interface AccountOptions {
  id: string;
  name?: string;
  rootEmail?: string;
  isManagementAccount?: boolean;
  organizationalUnit?: OrganizationalUnit;
}

export class Account extends Component {
  // member components
  public readonly regions: Region[];
  // parent
  public readonly organization: Organization;
  // all other options
  public readonly options: AccountOptions;

  constructor(organization: Organization, options: AccountOptions) {
    super(organization.project);

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components
    this.regions = [];

    // parents + inverse
    this.organization = organization;
    this.organization.accounts.push(this);

    // all other options
    const defaultOptions: Partial<AccountOptions> = {
      isManagementAccount: false,
    };
    this.options = { ...defaultOptions, ...options };

    // if OU defined, add account to OU
    if (this.options.organizationalUnit) {
      this.options.organizationalUnit.accounts.push(this);
    }

    // debugging output
    this.project.logger.info(`INIT Account: "${this.id}"`);
  }

  public get id() {
    return this.options.id;
  }

  public get name() {
    return this.options.name;
  }

  public get rootEmail() {
    return this.options.rootEmail;
  }

  public get isManagementAccount() {
    return this.options.isManagementAccount;
  }

  public get organizationalUnit() {
    return this.options.organizationalUnit;
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add an region to an account.
   *
   * @param {RegionOptions}
   * @returns {Region}
   */
  public addRegion(options: RegionOptions) {
    const region = this.regions.find((r) => r.id === options.id);
    return region ? region : new Region(this, options);
  }
}
