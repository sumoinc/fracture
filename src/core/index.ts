export * from "./access-pattern";
export * from "./account";
export * from "./audit-strategy";
export * from "./fracture";
export * from "./index";
export * from "./naming-strategy";
export * from "./operation";
export * from "./organization";
export * from "./organizational-unit";
export * from "./region";
export * from "./resource-attribute";
export * from "./resource";
export * from "./service";
export * from "./structure";
