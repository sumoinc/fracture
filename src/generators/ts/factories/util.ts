import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
