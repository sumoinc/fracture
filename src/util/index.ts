export * from "./determineBuildId";
export * from "./test-util";
