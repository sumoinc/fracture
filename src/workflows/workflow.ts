import { Component } from "projen";
import { GitHub, GithubWorkflow } from "projen/lib/github";
import { TypeScriptProject } from "projen/lib/typescript";
import { BuildJob } from "./jobs/build-job";
import { DeployJob, DeployJobOptions } from "./jobs/deploy-job";
import { WorkflowJob } from "./jobs/workflow-job";

/**
 * Name of the permission back up file to include in the build artifact
 * to work around a GitHub Action bug that does not preserve file mode
 * permissions across upload and download actions.
 *
 * See {@link https://github.com/actions/upload-artifact/issues/38}
 */
export const PERMISSION_BACKUP_FILE = "permissions-backup.acl";

export interface WorkflowOptions {
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   *
   * @default workflow
   */
  readonly name?: string;

  /**
   * The node version to use in workflows.
   *
   * @default latest
   */
  readonly nodeVersion?: string;

  /**
   * Enable Node.js package cache in GitHub workflows.
   *
   * @default true
   */
  readonly packageCache?: boolean;

  /**
   * The version of PNPM to use if using PNPM as a package manager.
   *
   * @default "8"
   */
  readonly pnpmVersion?: string;

  /**
   * Automatically update files modified during builds to pull-request branches.
   * This means that any files synthesized by projen or e.g. test snapshots will
   * always be up-to-date before a PR is merged.
   *
   * Implies that PR builds do not have anti-tamper checks.
   *
   * This is enabled by default only if `githubTokenSecret` is set. Otherwise it
   * is disabled, which implies that file changes that happen during build will
   * not be pushed back to the branch.
   *
   * @default false
   */
  readonly mutableBuild?: boolean;

  /**
   * What runners should this workflow use by default?
   *
   * @default ["ubuntu-latest"]
   */
  readonly defaultRunners?: Array<string>;
}

export class Workflow extends Component {
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   */
  public readonly name: string;

  /**
   * The node version to use in workflows.
   *
   * @default latest
   */
  public readonly nodeVersion: string;

  /**
   * Enable Node.js package cache in GitHub workflows.
   *
   * @default true
   */
  public readonly packageCache: boolean;

  /**
   * Branch or tag name.
   * @default - the default branch is implicitly used
   */
  public readonly ref?: string;

  /**
   * The repository (owner/repo) to use.
   * @default - the default repository is implicitly used
   */
  public readonly repository?: string;

  /**
   * Whether to download files from Git LFS for this workflow
   *
   * @default - Use the setting on the corresponding GitHub project
   */
  public readonly downloadLfs?: boolean;

  /**
   * The version of PNPM to use if using PNPM as a package manager.
   *
   * @default "8"
   */
  public readonly pnpmVersion: string;

  /**
   * Automatically update files modified during builds to pull-request branches.
   * This means that any files synthesized by projen or e.g. test snapshots will
   * always be up-to-date before a PR is merged.
   *
   * Implies that PR builds do not have anti-tamper checks.
   *
   * This is enabled by default only if `githubTokenSecret` is set. Otherwise it
   * is disabled, which implies that file changes that happen during build will
   * not be pushed back to the branch.
   *
   * @default false
   */
  readonly mutableBuild: boolean;

  /**
   * What runners should this workflow use by default?
   *
   * @default ["ubuntu-latest"]
   */
  readonly defaultRunners: Array<string>;

  /**
   * Instance of github configuration for the project
   */
  public readonly github: GitHub;

  /**
   * Github workflow Workflow
   */
  public readonly workflow: GithubWorkflow;

  /**
   * Build Job for the workflow.
   */
  public readonly buildJob: BuildJob;

  /**
   * Other Jobs in this workflow
   */
  public readonly otherJobs: Array<WorkflowJob> = [];

  constructor(
    public readonly project: TypeScriptProject,
    options: WorkflowOptions
  ) {
    super(project);

    const github = GitHub.of(project);

    if (!github) {
      throw new Error(
        "Workflows are currently only supported for GitHub projects"
      );
    }

    // defaults
    this.name = options.name ?? "workflow";
    this.nodeVersion = options.nodeVersion ?? "latest";
    this.packageCache = options.packageCache ?? true;
    this.pnpmVersion = options.pnpmVersion ?? "8";
    this.mutableBuild = options.mutableBuild ?? false;
    this.defaultRunners = options.defaultRunners ?? ["ubuntu-latest"];

    // local
    this.github = github;
    this.workflow = new GithubWorkflow(github, this.name);

    // render the build steps
    this.buildJob = new BuildJob(this);
  }

  addDeployJob(options: DeployJobOptions): DeployJob {
    const job = new DeployJob(this, { ...options });
    this.otherJobs.push(job);
    return job;
  }

  preSynthesize(): void {
    // add the build job to the workflow
    this.workflow.addJob(this.buildJob.jobId, this.buildJob.render());
    // add other jobs
    this.otherJobs.forEach((job) => {
      this.workflow.addJob(job.jobId, job.render());
    });
  }
}
