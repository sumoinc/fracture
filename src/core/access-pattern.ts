import { deepMerge } from "projen/lib/util";
import { ValueOf } from "type-fest";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import {
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
} from "./resource-attribute";
import { FractureComponent } from "../core/component";
import { DynamoGsi } from "../dynamodb/dynamo-gsi";
import { DynamoTable } from "../dynamodb/dynamo-table";

export const ACCESS_PATTERN_TYPE = {
  /**
   * Main PK and PK pattern used to identify this record.
   */
  IDENTIFIER: "Identifier",
  /**
   * Lookup pattern for quick begins_with() searching and such.
   */
  LOOKUP: "lookup",
  /**
   * Any other AP
   */
  OTHER: "other",
} as const;

export interface AccessPatternOptions {
  dynamoGsi?: DynamoGsi;
  pkAttributeOptions?: ResourceAttributeOptions;
  skAttributeOptions?: ResourceAttributeOptions;
  type?: ValueOf<typeof ACCESS_PATTERN_TYPE>;
}

export class AccessPattern extends FractureComponent {
  // member components
  public pkAttribute: ResourceAttribute;
  public skAttribute: ResourceAttribute;
  public operations: Operation[] = [];
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: Required<AccessPatternOptions>;

  constructor(resource: Resource, options: AccessPatternOptions) {
    super(resource.fracture);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     **************************************************************************/

    // make a new gsi is one wasn't passed into the Access Pattern
    const dynamoGsi = options.dynamoGsi
      ? options.dynamoGsi
      : new DynamoGsi(resource.dynamoTable);

    const defaultOptions: Required<AccessPatternOptions> = {
      dynamoGsi,
      pkAttributeOptions: {
        name: dynamoGsi.pkName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      skAttributeOptions: {
        name: dynamoGsi.skName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      type: ACCESS_PATTERN_TYPE.OTHER,
    };

    /***************************************************************************
     *
     * INIT ACCESS PATTERN
     *
     **************************************************************************/

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<AccessPatternOptions>;
    // the attribute might already be defined.
    const pkAttribute = resource.getAttributeByName(dynamoGsi.pkName);
    const skAttribute = resource.getAttributeByName(dynamoGsi.skName);

    // member components
    this.pkAttribute = pkAttribute
      ? pkAttribute
      : new ResourceAttribute(resource, this.options.pkAttributeOptions);
    this.skAttribute = skAttribute
      ? skAttribute
      : new ResourceAttribute(resource, this.options.skAttributeOptions);

    // parent + inverse
    this.resource = resource;
    this.resource.accessPatterns.push(this);

    this.project.logger.info(
      `Access Pattern for: "${this.pkAttribute.name}:${this.skAttribute.name}" initialized.`
    );

    return this;
  }

  public build() {
    this.project.logger.debug(
      `BUILD Access Pattern: "${this.pkAttribute.name}:${this.skAttribute.name}" called.`
    );
    /***************************************************************************
     *
     * CRUD OPERATIONS
     *
     * TODO
     *
     **************************************************************************/
    /*
    new Operation(this, {
      operationType: OPERATION_TYPE.MUTATION,
      operationSubType: OPERATION_SUB_TYPE.LIST,
    });
    */
  }

  addPkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.pkAttribute.compositionSources.push(sourceAttribute);
  }

  addSkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.skAttribute.compositionSources.push(sourceAttribute);
  }

  /**
   * Returns operation for given sub-type
   */
  public getOperation(
    operationSubType: ValueOf<typeof OPERATION_SUB_TYPE>
  ): Operation {
    const returnOperation = this.operations.find(
      (o) => o.operationSubType === operationSubType
    );

    if (!returnOperation) {
      throw new Error(`Operation not found for sub-type: ${operationSubType}`);
    }

    return returnOperation;
  }

  public get createOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.CREATE_ONE);
  }

  public get readOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.READ_ONE);
  }

  public get updateOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.UPDATE_ONE);
  }

  public get deleteOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.DELETE_ONE);
  }

  public get isKeyAccessPattern(): boolean {
    return this.dynamoTable.keyDynamoGsi === this.dynamoGsi;
  }

  public get isLookupAccessPattern(): boolean {
    return this.dynamoTable.lookupDynamoGsi === this.dynamoGsi;
  }

  public get dynamoTable(): DynamoTable {
    return this.dynamoGsi.dynamoTable;
  }

  public get dynamoGsi(): DynamoGsi {
    return this.options.dynamoGsi;
  }
}
