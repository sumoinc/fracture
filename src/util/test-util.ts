import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";

export const TEST_ACCOUNT_ONE = "000000000000";
export const TEST_ORG_ONE = "org-123456";
export const TEST_REGION_ONE = "us-east-1";

/**
 *
 * Helper to generate and return one file from fracture.
 * Useful when trying to test the final output of generated files.
 *
 * @param fracture
 * @param filepath
 * @returns collection of files starting with supplied filepath
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered;
};

export const synthFile = (project: Project, filepath: string): string => {
  const files = synthFiles(project, filepath);

  if (files.length > 1) {
    throw new Error(`more than one file matched filepath "${filepath}"`);
  }

  if (files.length === 0) {
    throw new Error(`No files matched filepath "${filepath}"`);
  }

  return files[filepath];
};

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testProject = () => {
//   return new Project({
//     name: "my-project",
//   });
// };

// export const testNodeProject = (options: Partial<NodeProjectOptions> = {}) => {
//   return new NodeProject({
//     name: "my-project",
//     defaultReleaseBranch: "main",
//     packageManager: NodePackageManager.PNPM,
//     pnpmVersion: "8",
//     ...options,
//   });
// };

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testFractureProject = (
//   options: Partial<FractureProjectOptions> = {}
// ) => {
//   const fp = new FractureProject({
//     name: "my-project",
//     ...options,
//   });

//   /*
//   fp.buildEnvironment = new AwsEnvironment(fp, {
//     name: "test",
//     account: TEST_ACCOUNT_ONE,
//     region: TEST_REGION_ONE,
//     primaryHostedZoneId: "foo",
//     primaryDomainName: "bar",
//   });
//   */

//   return fp;
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testApp = () => {
//   return new App({
//     parent: testFractureProject(),
//     name: "my-app",
//   });
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testDataService = () => {
//   return new DataService(testApp(), {
//     name: "my-service",
//   });
// };

// /**
//  * Build a simple resource to test with
//  */
// export const testResource = () => {
//   const service = testDataService();
//   return new Resource(service.app, {
//     service,
//     name: "my-resource",
//   });
// };

// /**
//  * A more complicated resource configuration
//  */
// export const complexService = () => {
//   const service = testDataService();

//   const person = new Resource(service.app, { service, name: "person" });

//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     comments: ["A club"],
//     attributeOptions: [{ name: "name" }],
//   });

//   person.addAttribute({
//     name: "first-name",
//     shortName: "fn",
//   });
//   person.addAttribute({
//     name: "last-name",
//     shortName: "ln",
//   });
//   person.addAttribute({
//     name: "height",
//     shortName: "ht",
//     type: ResourceAttributeType.INT,
//   });
//   person.addAttribute({
//     name: "is-frog",
//     shortName: "if",
//     type: ResourceAttributeType.BOOLEAN,
//     comments: ["Is this person a frog?"],
//   });
//   person.addAttribute({
//     name: "fav-colors",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: ResourceAttributeType.STRING,
//     comments: ["List opf favorite colors"],
//   });

//   // one specific club
//   person.addAttribute({
//     name: "one-club",
//     type: club,
//   });
//   // an array of clubs
//   person.addAttribute({
//     name: "array-of-clubs",
//     shortName: "ad",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });

//   // an array of persons
//   club.addAttribute({
//     name: "array-of-persons",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });

//   // operation
//   Operation.create(person);
//   Operation.read(person);
//   Operation.update(person);
//   Operation.delete(person);

//   return service;
// };

// /**
//  * Build a simple resource attribute to test with
//  */
// export const testResourceAttribute = () => {
//   const service = testDataService();
//   const resource = testResource();
//   return new ResourceAttribute(service.app, {
//     name: "my-attr",
//     resource,
//   });
// };

// /**
//  * Build a simple structure to test with
//  */
// export const testStructure = () => {
//   const service = testDataService();
//   return new Structure(service.app, { name: "my-structure" });
// };

// /**
//  * Build a simple operation to test with
//  */
// export const testOperation = () => {
//   const resource = testResource();
//   return new Operation(resource.service.app, {
//     name: "my-operation",
//     resource,
//   });
// };

// /**
//  * Build a nuxt site to test with
//  */
// export const testNuxtJsSite = () => {
//   return new NuxtJsSite({
//     parent: testFractureProject(),
//     name: "my-site",
//   });
// };
// export const testServiceTwoRelatedResources = () => {
//   const service = testDataService();
//   const person = new Resource(service.app, {
//     service,
//     name: "person",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   club.addAttribute({
//     name: "people",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });
//   person.addAttribute({
//     name: "clubs",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });
//   return { service, person, club };
// };
