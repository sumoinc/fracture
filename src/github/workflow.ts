import { paramCase } from "change-case";
import {
  GitHub,
  GithubWorkflow,
  GithubWorkflowOptions,
} from "projen/lib/github";
import { BuildJob } from "./build-job";
import { Fracture } from "../core";
import { TurboRepo } from "../turborepo";

export interface GithubFractureWorkflowOptions extends GithubWorkflowOptions {
  /**
   * The name of the workflow.
   */
  name: string;
  /**
   * What branches shoudl trigger this pipeline?
   */
  branchTriggerPatterns: string[];
  /**
   * What paths should trigger this pipeline?
   */
  pathTriggerPatterns: string[];
  /**
   * Pipelines always builds but don't always deploy.
   * Should this pipeline deploy the app after build?
   */
  deploy: boolean;
}

export class GithubFractureWorkflow extends GithubWorkflow {
  /**
   * The name of the workflow.
   */
  public readonly name: string;
  /**
   * Storage location for this workflow.
   */
  public readonly filePath: string;

  constructor(fracture: Fracture, options: GithubFractureWorkflowOptions) {
    /***************************************************************************
     * Super
     **************************************************************************/

    const github = GitHub.of(fracture);
    if (!github) {
      throw new Error(
        "Fracture workflows are currently only supported for GitHub enabled projects"
      );
    }
    const name = options.name ? paramCase(options.name) : "deployment";
    super(github, name, options);

    /***************************************************************************
     * Locals
     **************************************************************************/

    const turboRepo = TurboRepo.of(fracture);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.name = name;
    this.filePath = `.github/workflows/${this.name}.yml`;

    /***************************************************************************
     * Triggers
     **************************************************************************/

    this.on({
      push: {
        branches: options.branchTriggerPatterns,
        paths: options.pathTriggerPatterns,
      },
      workflowDispatch: {}, // allow manual triggering
    });

    /***************************************************************************
     *
     * BUILD JOB
     *
     * Leverage turbo-repo commands, if available.
     *
     **************************************************************************/

    const buildJob = !turboRepo
      ? new BuildJob(fracture)
      : new BuildJob(fracture, {
          buildCommand: fracture.runTaskCommand(turboRepo.buildTask),
          synthCommand: fracture.runTaskCommand(turboRepo.synthTask),
        });
    this.addJob(buildJob.jobId, buildJob.job);

    /***************************************************************************
     *
     * DEPLOY JOBS
     *
     **************************************************************************/

    // no waves, just use default workflow
    /*
    if (pipeline.waves.length === 0) {
      const deployJob = !turboRepo
        ? new DeployJob(fracture)
        : new DeployJob(fracture, {
            deployCommand: fracture.runTaskCommand(turboRepo.deployTask),
          });
      this.githubWorkflow.addJob(deployJob.jobId, deployJob.job);
    }*/

    // add deployment waves to workflow
    /*
    if (pipeline.waves.length > 0) {
      pipeline.waves.forEach((wave) => {
        wave.stages.forEach(() => {
          const deployJob = new DeployJob(pipeline.project as Fracture);
          this.githubWorkflow.addJob(deployJob.jobId, deployJob.job);
          //console.log(wave.name, stage.name);
          //this.addDeployJob(stage);
        });
      });
    }
    */
  }
}
