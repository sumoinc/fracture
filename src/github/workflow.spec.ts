import { GithubFractureWorkflow } from "./workflow";
import { Fracture } from "../core";
import { synthFile } from "../util/test-util";

let fracture: Fracture;

beforeEach(() => {
  fracture = new Fracture();
});

/*******************************************************************************
 * DEFAULT WORKFLOWS
 ******************************************************************************/

describe("Default workflow", () => {
  test("no args", () => {
    const workflow = new GithubFractureWorkflow(fracture, {
      name: "foo",
      branchTriggerPatterns: ["main"],
      pathTriggerPatterns: ["src/**/*"],
      deploy: true,
    });

    const content = synthFile(fracture, workflow.filePath);
    expect(content).toMatchSnapshot();
    console.log(content);
  });

  test("without turborepo", () => {
    const noTurbo = new Fracture({
      turboRepoEnabled: false,
    });
    const workflow = new GithubFractureWorkflow(noTurbo, {
      name: "foo",
      branchTriggerPatterns: ["main"],
      pathTriggerPatterns: ["src/**/*"],
      deploy: true,
    });

    const content = synthFile(noTurbo, workflow.filePath);
    expect(content).toMatchSnapshot();
    //console.log(content);
  });
});

/*******************************************************************************
 * WAVES & STAGES
 ******************************************************************************/

/*
describe.only("Waves & Stages", () => {
  test("one wave, one stage", () => {
    const fracture = new Fracture({
      name: "test-project",
      logging: {
        level: LogLevel.OFF,
      },
    });
    const pipeline = new Pipeline(fracture, {
      name: "default",
      branchTriggerPatterns: ["main"],
    });
    const workflow = new GithubFractureWorkflow(fracture, { pipeline });

    const content = synthWorkflow(workflow);
    expect(content).toMatchSnapshot();
    console.log(content);
  });
});
*/
/*******************************************************************************
 * MISC TESTS
 ******************************************************************************/
/*
test("has valid filepath", () => {
  const fracture = new Fracture({
    name: "test-project",
    logging: {
      level: LogLevel.OFF,
    },
  });
  const pipeline = new Pipeline(fracture, {
    name: "default",
    branchTriggerPatterns: ["main"],
  });
  const workflow = new GithubFractureWorkflow(fracture, {
    name: "Foo Bar Baz",
    pipeline,
  });

  expect(workflow.filePath).toBe(".github/workflows/foo-bar-baz.yml");
});
*/
/*
describe("multiple waves", () => {
  test("Workflows depend on build", () => {
    const pipeline = app.addPipeline({
      name: "default",
      branchTriggerPatterns: ["main"],
    });
    pipeline.addStage({ name: "East Coast", environment: devUsEast });
    pipeline.addStage({ name: "West Coast", environment: devUsWest });
    const workflow = new GithubFractureWorkflow(fracture, { pipeline });

    const content = synthWorkflow(workflow);
    expect(content).toMatchSnapshot();

    console.log(content);
  });
});
*/

/**
 *
 * Helper function to generate a workflow file so we can test it's structure.
 *
 */
/*
function synthWorkflow(workflow: Workflow): any {
  const snapshot = synthSnapshot(workflow.project);
  const filtered = Object.keys(snapshot)
    .filter((path) => path.match(workflow.filePath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered[workflow.filePath];
}*/
