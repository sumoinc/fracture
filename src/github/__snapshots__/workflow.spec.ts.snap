// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default workflow no args 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: foo
on:
  push:
    branches:
      - main
    paths:
      - src/**/*
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: \\"8\\"
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: \\"18\\"
          cache: pnpm
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: Build
        run: npx projen turbo:build
      - name: Synth
        run: npx projen turbo:synth
"
`;

exports[`Default workflow without turborepo 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: foo
on:
  push:
    branches:
      - main
    paths:
      - src/**/*
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: \\"8\\"
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: \\"18\\"
          cache: pnpm
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: Build
        run: npx projen build
      - name: Synth
        run: npx projen synth
"
`;
